target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c

```diff
@@ -0,0 +1,484 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/module.h>
+#include <linux/of_platform.h>
+#include "i2c-rtl9300.h"
+
+#define REG(x)		(i2c->base + x + (i2c->scl_num ? 0x1c : 0))
+#define REG_MASK(clear, set, reg)	\
+			writel((readl(REG(reg)) & ~(clear)) | (set), REG(reg))
+
+struct i2c_drv_data {
+	int scl0_pin;
+	int scl1_pin;
+	int sda0_pin;
+	struct i2c_algorithm *algo;
+	int (*read)(struct rtl9300_i2c *i2c, u8 *buf, int len);
+	int (*write)(struct rtl9300_i2c *i2c, u8 *buf, int len);
+	void (*reg_addr_set)(struct rtl9300_i2c *i2c, u32 reg, u16 len);
+	int (*config_xfer)(struct rtl9300_i2c *i2c, u16 addr, u16 len);
+	int (*execute_xfer)(struct rtl9300_i2c *i2c, char read_write, int size,
+			    union i2c_smbus_data * data, int len);
+	void (*writel)(struct rtl9300_i2c *i2c, u32 data);
+	void (*config_io)(struct rtl9300_i2c *i2c, int scl_num, int sda_num);
+};
+			
+DEFINE_MUTEX(i2c_lock);
+
+static void rtl9300_i2c_reg_addr_set(struct rtl9300_i2c *i2c, u32 reg, u16 len)
+{
+	// Set register address width
+	REG_MASK(0x3 << RTL9300_I2C_CTRL2_MADDR_WIDTH, len << RTL9300_I2C_CTRL2_MADDR_WIDTH,
+		 RTL9300_I2C_CTRL2);
+
+	// Set register address
+	REG_MASK(0xffffff << RTL9300_I2C_CTRL1_MEM_ADDR, reg << RTL9300_I2C_CTRL1_MEM_ADDR,
+		 RTL9300_I2C_CTRL1);
+}
+
+static void rtl9310_i2c_reg_addr_set(struct rtl9300_i2c *i2c, u32 reg, u16 len)
+{
+	// Set register address width
+	REG_MASK(0x3 << RTL9310_I2C_CTRL_MADDR_WIDTH, len << RTL9310_I2C_CTRL_MADDR_WIDTH,
+		 RTL9310_I2C_CTRL);
+
+	// Set register address
+	writel(reg, REG(RTL9310_I2C_MEMADDR));
+}
+
+static void rtl9300_i2c_config_io(struct rtl9300_i2c *i2c, int scl_num, int sda_num)
+{
+	u32 v;
+
+	// Set SCL pin
+	REG_MASK(0, BIT(RTL9300_I2C_CTRL1_GPIO8_SCL_SEL), RTL9300_I2C_CTRL1);
+
+	// Set SDA pin
+	REG_MASK(0x7 << RTL9300_I2C_CTRL1_SDA_OUT_SEL,
+		 i2c->sda_num << RTL9300_I2C_CTRL1_SDA_OUT_SEL, RTL9300_I2C_CTRL1);
+
+	// Set SDA pin to I2C functionality
+	v = readl(i2c->base + RTL9300_I2C_MST_GLB_CTRL);
+	v |= BIT(i2c->sda_num);
+	writel(v, i2c->base + RTL9300_I2C_MST_GLB_CTRL);
+}
+
+static void rtl9310_i2c_config_io(struct rtl9300_i2c *i2c, int scl_num, int sda_num)
```

Comment:
This appears to have the same logic as rtl9300_i2c_config_io, just with different offsets. Can't these be stored in in e.g. rtl9300_i2c, so the same logic can be used?

Reply:
You are right. But this seems to be a matter of taste. I was using stored offsets extensively in the dsa/ethernet drivers. But people like Sebastian go after them and prefer to have duplicate code called via function pointers like here. The argument seems to be that then you don't see the offset values directly in the code, which is important to verify that it works correctly. Forgetting to set one of the offset values in the structures is a very good way of coding a very difficult to find bug. We were using a redefined REG() macro to find all the instances for the RTL931X, otherwise it is not possible if you have hundreds of offsets on 4 different SoC architectures. The good thing about forgetting to set function pointers is the rather dramatic and immediate crash.

Reply:
Copy-pasting code is also a very good way of creating bugs. Again, I think this only makes the code harder to understand when another person is reading it. This style implements a certain 'device op' for RTL9310 and RTL9300. Most of these are rather shot, but the execute_transfer() op is quite a bit more complicated. If both I2C controllers work in (exactly) the same way, there should only be one implementation of that logic in my opinion.

Reply:
I see now that the registers differ between RTL9300 and RTL9310. They just happen to have the same control bits laid out differently over the two control registers :expressionless:
I still think there should be a better way of abstracting this. With a regmap interface, you could define fields for all the control bits, and put these in a private data structure. Then you would only need one function to implement each piece of logic, and could let those refer to the regmap fields.